### ISSUE TITLE
v8::AdjustAmountOfExternalAllocatedMemory()

### ISSUE THREAD

**Issue by springmeyer** [mapbox/carmen-cache#76]
We should try estimating the memory usage of carmen cache and setting `v8::AdjustAmountOfExternalAllocatedMemory()`. This should help with scenarios where:

 - huge caches are loaded quickly into the C++ heap
 - v8 needs to garbage collect its JS objects to keep memory down...
 - But v8 may decide not to garbage collect since it does not see the C++ heap and thinks it has ample available memory

Using `v8::AdjustAmountOfExternalAllocatedMemory()` will likely slow down the performance of repeated calls to `cache.loadSync` but may allow node to avoid using too much memory for temporary objects created in the same code that is calling `cache.loadSync`.

refs https://github.com/nodejs/nan/blob/master/doc/v8_internals.md#nanadjustexternalmemory

**Comment by springmeyer** [comment]
To know how to roughly estimate memory we need to know the footprint of each C++ member of `Carmen::Cache`. A quick hack to get these is:

```diff
 extern "C" {
     static void start(Handle<Object> target) {
+        std::clog << "Cache::intarray " << sizeof(Cache::intarray) << "\n";
+        std::clog << "Cache::arraycache " << sizeof(Cache::arraycache) << "\n";
+        std::clog << "Cache::memcache " << sizeof(Cache::memcache) << "\n";
+        std::clog << "Cache::message_pair " << sizeof(Cache::message_pair) << "\n";
+        std::clog << "Cache::message_list " << sizeof(Cache::message_list) << "\n";
+        std::clog << "Cache::message_cache " << sizeof(Cache::message_cache) << "\n";
         Cache::Initialize(target);
     }
 }
```

Then:

```bash
$ ./node_modules/.bin/node-pre-gyp rebuild
$ node -e "require('./lib/carmen.node')"
Cache::intarray 24
Cache::arraycache 24
Cache::memcache 24
Cache::message_pair 48
Cache::message_list 24
Cache::message_cache 24
```

**Comment by springmeyer** [comment]
Going to close this. The idea still stands: that telling v8 about external memory might help avoid running low by triggering v8 to garbage collect more often. However the original OOM that we saw which prompted me to create this issue was fully solved by #82. So, this does not feel necessary to move on at this time.
